agent/ecr/client_test.go

```diff
@@ -0,0 +1,183 @@
+			RegistryIds: []*string{aws.String(testRegistryId)},
+		}).Return(&ecrapi.GetAuthorizationTokenOutput{
+		AuthorizationData: []*ecrapi.AuthorizationData{testAuthData},
+	}, nil)
+
+	suite.mockCache.EXPECT().Get(testRegistryId).Return(nil, false)
+	suite.mockCache.EXPECT().Set(testRegistryId, testAuthData)
+
+	authorizationData, err := suite.ecrClient.GetAuthorizationToken(testRegistryId)
+	assert.NoError(suite.T(), err)
+	assert.Equal(suite.T(), testAuthData, authorizationData)
+}
+
+func (suite *GetAuthorizationTokenTestSuite) TestGetAuthorizationTokenCacheHitExpiredToken() {
+	expiredAuthData := &ecrapi.AuthorizationData{
+		ProxyEndpoint:      aws.String(testProxyEndpoint),
+		AuthorizationToken: aws.String(testToken),
+		ExpiresAt:          aws.Time(time.Now().Add(-5 * time.Minute)),
+	}
+
+	testAuthData := &ecrapi.AuthorizationData{
+		ProxyEndpoint:      aws.String(testProxyEndpoint),
+		AuthorizationToken: aws.String(testToken),
+		ExpiresAt:          aws.Time(time.Now().Add(12 * time.Hour)),
+	}
+
+	suite.mockClient.EXPECT().GetAuthorizationToken(
+		&ecrapi.GetAuthorizationTokenInput{
+			RegistryIds: []*string{aws.String(testRegistryId)},
+		}).Return(&ecrapi.GetAuthorizationTokenOutput{
+		AuthorizationData: []*ecrapi.AuthorizationData{testAuthData},
+	}, nil)
+
+	suite.mockCache.EXPECT().Get(testRegistryId).Return(expiredAuthData, true)
+	suite.mockCache.EXPECT().Set(testRegistryId, testAuthData)
+
+	authorizationData, err := suite.ecrClient.GetAuthorizationToken(testRegistryId)
+	assert.NoError(suite.T(), err)
+	assert.Equal(suite.T(), testAuthData, authorizationData)
+}
+
+func (suite *GetAuthorizationTokenTestSuite) TestGetAuthorizationTokenMissingAuthData() {
+	suite.mockClient.EXPECT().GetAuthorizationToken(
+		&ecrapi.GetAuthorizationTokenInput{
+			RegistryIds: []*string{aws.String(testRegistryId)},
+		}).Return(&ecrapi.GetAuthorizationTokenOutput{
+		AuthorizationData: []*ecrapi.AuthorizationData{},
+	}, nil)
+
+	suite.mockCache.EXPECT().Get(testRegistryId).Return(nil, false)
+
+	authorizationData, err := suite.ecrClient.GetAuthorizationToken(testRegistryId)
+	assert.Error(suite.T(), err)
+	assert.Nil(suite.T(), authorizationData)
+}
+
+func (suite *GetAuthorizationTokenTestSuite) TestGetAuthorizationTokenError() {
+	suite.mockClient.EXPECT().GetAuthorizationToken(
+		&ecrapi.GetAuthorizationTokenInput{
+			RegistryIds: []*string{aws.String(testRegistryId)},
+		}).Return(nil, errors.New("Nope Nope Nope"))
+
+	suite.mockCache.EXPECT().Get(testRegistryId).Return(nil, false)
+
+	authorizationData, err := suite.ecrClient.GetAuthorizationToken(testRegistryId)
+	assert.Error(suite.T(), err)
+	assert.Nil(suite.T(), authorizationData)
+}
+
+func (suite *GetAuthorizationTokenTestSuite) TestIsTokenValid() {
+
+	var testAuthTimes = []struct {
+		expireIn time.Duration
+		expected bool
+	}{
+		{-1 * time.Minute, false},
+		{time.Duration(0), false},
+		{1 * time.Minute, false},
+		{30 * time.Minute, false},
```

Comment:
You can use the `minimumExpirationJitter` and `maximumExpirationJitter` constants here.